stage1/init/init.go

```diff
@@ -396,6 +586,70 @@ func stage1() int {
 	return 0
 }
 
+func getUnitFileName() (unit string, err error) {
+	handle := C.dlopen(C.CString("libsystemd.so"), C.RTLD_LAZY)
+	if handle == nil {
+		err = fmt.Errorf("error opening libsystemd.so")
+		return
+	}
+	defer func() {
+		if r := C.dlclose(handle); r != 0 {
+			err = fmt.Errorf("error closing libsystemd.so")
+		}
+	}()
+
+	sd_pid_get_unit := C.dlsym(handle, C.CString("sd_pid_get_unit"))
+	if sd_pid_get_unit == nil {
+		err = fmt.Errorf("error resolving sd_pid_get_unit function")
+		return
+	}
+
+	var s string
+	u := C.CString(s)
+	defer C.free(unsafe.Pointer(u))
+
+	ret := C.my_sd_pid_get_unit(sd_pid_get_unit, 0, &u)
+	if ret < 0 {
+		err = fmt.Errorf("error calling sd_pid_get_unit: %v", syscall.Errno(-ret))
+		return
+	}
+
+	unit = C.GoString(u)
+	return
+}
+
+func getSlice() (slice string, err error) {
+	handle := C.dlopen(C.CString("libsystemd.so"), C.RTLD_LAZY)
```

Comment:
getSlice() is expensive because it calls dlopen() and dlclose() on each call. getSlice() is called in a loop by createCgroups(). But the returned value is not going to change.

Can you cache the value, either here or in createCgroups()?